# Cleaning Up Data from eBay Kleinanzeige
 - This data is from eBay Kleinanzeigen - a classified section of the German eBay website
 - In this notebook, the program will clean the data and analyze the data

## Part I - Introduction

In [1]:
import pandas as pd
import numpy as np
autos=pd.read_csv('autos.csv', encoding='Latin-1')

In [2]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine_

In [3]:
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

### Observations
 - Mix of objects and integer columns
 - Not all columns have 5,000 non-null values meaning some are nulls


## Part II - Cleaning Column Names

In [4]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [5]:
copy_autos=autos.copy()
print(copy_autos.head())
type(copy_autos)
copy_autos.columns

           dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere 

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [6]:
copy_autos.rename(columns = {'yearOfRegistration':'registration_year',
                                       'monthOfRegistration':'registration_month',
                                       'notRepairedDamage':'unrepaired_damage',
                                       'dateCreated':'ad_created',
                                       'dateCrawled':'date_crawled',
                                       'offerType':'offer_type',
                                       'abtest':'ab_test',
                                       'vehicleType':'vehicle_type',
                                       'fuelType':'fuel_type',
                                       'nrOfPictures':'nr_pf_pictures',
                                       'postalCode':'postal_code',
                                       'lastSeen':'last_seen'},inplace=True)


In [7]:
copy_autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'powerPS', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_pf_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

In [8]:
autos.columns=copy_autos.columns
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  powerPS   model   odometer  registration_month fuel_type  \
0    manuell      158  andere  150,000km                   3       lpg   
1  automatik      286     7er  150,000km                   6    benzin   
2    manuell      102    golf   70,000km                   7    benzin   
3  automatik       71  fortwo   70,000km                   6    benzin   
4    manuell        0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_pf_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

- Changes were made in column names from CamelCase to snake_case so it would be easier to replace spaces with underscores and for an easier read

## Part III Initial Exploration and Cleaning

In [9]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-04-04 16:40:33  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year  gearbox       powerPS  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_pf_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

- Columns with 2 unique options: seller, offer_type, abtest, gearbox, unrepaired_damage
- Columns that need more investigation: ab_test to see what a/b test was conducted and thos with only 2 unique options to investigate whether keeping those columns is essential to this investigation
- columns that store numeric data as text:  price, odometer (dates and timestamps are also not numeric, but in date/time formated i.e. date_crawled, ad_created and last_seen)


In [10]:
autos['price']=autos['price'].str.replace('$','').str.replace(',','')
autos['odometer']=autos['odometer'].str.replace('km','').str.replace(',','')
autos.head()
autos['price']=autos['price'].astype(int)
autos['odometer']=autos['odometer'].astype(int)
print(autos['price'].dtype)
print(autos['odometer'].dtype)
autos.rename(columns={'odometer':'odometer_km'},inplace=True)

int64
int64


In [11]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type  price  ab_test vehicle_type  registration_year  \
0  privat    Angebot   5000  control          bus               2004   
1  privat    Angebot   8500  control    limousine               1997   
2  privat    Angebot   8990     test    limousine               2009   
3  privat    Angebot   4350  control   kleinwagen               2007   
4  privat    Angebot   1350     test        kombi               2003   

     gearbox  powerPS   model  odometer_km  registration_month fuel_type  \
0    manuell      158  andere       150000                   3       lpg   
1  automatik      286     7er       150000                   6    benzin   
2    manuell      102    golf        70000                   7    benzin   
3  automatik       71  fortwo        70000                   6    benzin   
4    manuell        0   focus       150000                   7    benzin   

        brand unrepaired_damage           ad_created  nr_pf_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

## Part IV - Exploring the Odometer and Price Columns

In [12]:
print('\n number of values ') 
print(autos.shape[0])
print('\n number of unique values ') 
print(autos['odometer_km'].unique().shape)
print('\n the unique values ')
print(autos['odometer_km'].unique())
print('\n min,max,median,mean ')
print(autos['odometer_km'].describe())
print('\n descending order of top 50 ')
print(autos['odometer_km'].head(50).value_counts().sort_index(ascending=False))


 number of values 
50000

 number of unique values 
(13,)

 the unique values 
[150000  70000  50000  80000  10000  30000 125000  90000  20000  60000
   5000 100000  40000]

 min,max,median,mean 
count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

 descending order of top 50 
150000    38
125000     2
90000      1
80000      1
70000      2
60000      2
50000      1
30000      1
20000      1
10000      1
Name: odometer_km, dtype: int64


- The odometer values seem to be reasonable - of 5,000 values there are only 13 unique values that are multiples of 1,000 which lead to the belief that the odometer classification are ranges as oppose to absolute values
- The values range from 5,000 km to 150,000 km which is reasonable
- There seems to be no outliers in this dataset

In [13]:
print('\n number of values ') 
print(autos.shape[0])
print('\n number of unique values ') 
print(autos['price'].unique().shape)
print('\n the unique values ')
print(autos['price'].unique())
print('\n min,max,median,mean ')
print(autos['price'].describe())
print('\n descending order of top 50 ')
print(autos['price'].head(50).value_counts().sort_index(ascending=False))


 number of values 
50000

 number of unique values 
(2357,)

 the unique values 
[ 5000  8500  8990 ...   385 22200 16995]

 min,max,median,mean 
count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

 descending order of top 50 
56000    1
48500    1
41500    1
25450    1
18900    1
13800    1
11900    1
11500    1
9000     1
8990     1
8500     1
7999     1
7900     1
7500     1
5500     1
5299     1
5250     1
5000     1
4999     1
4800     1
4350     1
4150     1
3999     1
3500     2
3499     1
3200     1
2850     2
2600     1
1990     1
1450     1
1350     2
1200     1
1099     1
1000     1
999      1
900      1
777      1
590      1
350      3
300      2
250      1
90       1
80       1
0        1
Name: price, dtype: int64


- Of the 50,000 values in the price column there are 2,367 unique values meaning that either prices appear multiple times or there are some prices that repeat very often
- Looking at the sorted value count of the first 50 values, it seems that it is the former in that prices appear a multiple times
- The describe function gives the min and max which are quite suspicious as they are 0 and 1 million respectively. A further investigation needs to be conducted for this dataset
- One way to do so is to look at all the values below 2,000 and above 100,000
- **Examinated values below 2,000 and there were almost 20,000 therefore going to narrow it down to values below 500 - there were 5,000 values below 500, now going to look at below 100 - these values seem appropriate as they either have an old registration year or have a high odometer value or both - now going to examine under 50 which has about 1,600 values mostly 0's and 1's - going to see how many are below 1 - a little over 1400 values - going to eliminate values below 1 as starting selling a car for no money seems suspicious**
- **Examinated values above 100,000 - 53 values - some seem reasonable as they are expensive brands (i.e. Porsche) going to raise the value to 1,000,000 - 11 values - some models like the Ferrari's seem reasonable and some do not - time to research each car to see if the prices are reasonable - after researching some of the most expensive cars sold on auction as well as the models of the cars that make up these eleven values it seemed reasonable that values below 10,000 would be included as some prices such as 99,999,999 and 12,345,678 seemed to be suspicious from the sequencing of the numbers**

In [14]:
examlow=autos[autos['price']<1]
print(examlow.shape[0])
examlow

1421


date_crawled                                               name  \
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
71     2016-03-28 19:39:35  Suche_Opel_Astra_F__Corsa_oder_Kadett_E_mit_Re...   
80     2016-03-09 15:57:57  Nissan_Primera_Hatchback_1_6_16v_73_Kw___99Ps_...   
87     2016-03-29 23:37:22                      Bmw_520_e39_zum_ausschlachten   
99     2016-04-05 09:48:54                    Peugeot_207_CC___Cabrio_Bj_2011   
118    2016-03-12 05:03:00  VW_Sharan_V6_204_PS_Karosse_Rohkarosse_mit_Pap...   
146    2016-03-22 23:59:28                                    Ford_Fiesta_rot   
167    2016-04-02 19:43:45  Suche_VW_Multivan_Innenausstattung_Set_oder_TE...   
180    2016-03-19 10:50:25                                       Zu_verkaufen   
226    2016-03-25 23:52:12                      Porsche_911_S_Targa__67er_SWB   
234    2016-03-05 16:56:22                                   Fiat_Punto_Sport   
248    2016-03-08 22:57:53                             VW_Passat_zu_verkaufen   
259    2016-04-03 23:49:58                      guenstiges_Auto_/_auch_defekt   
301    2016-03-08 20:37:59                         Kaufe_alle_Autos_bietet_an   
323    2016-03-15 16:54:59                                W220_Unfallfahrzeug   
327    2016-04-03 13:49:21              BMW_Cabrio_318i_/_BMW_318_tds_Touring   
347    2016-04-02 11:48:02                                  Volvo_TP21__Sugga   
366    2016-03-21 20:36:25  Ford_Escort_Express_75__LKW_Zulussung___Diesel...   
389    2016-03-08 11:38:52                                    Vw_polo_classic   
418    2016-03-29 14:43:24         Fiat_SCUDO_8_Sitzer_Bus__Diesel_JTD__80_KW   
427    2016-03-12 11:44:47                 TOP_ZUSTAND_GrunePlakette_Euro4!!!   
430    2016-03-18 23:52:40                                  Winterraeder_FORD   
449    2016-04-03 20:59:59                           Tausche_gegen_ein_smart.   
487    2016-03-14 21:41:36  MAG_MICH_KEINER_HABEN_Opel_Omega_B_Caravan_2.0...   
502    2016-03-25 23:55:24                               C32_AMG_C55_C43_w203   
519    2016-03-28 08:53:58  Verkaufe/Tausche_Seat_Inca__VW_Caddy__aus_2002...   
565    2016-03-20 20:49:36     Tausche_Audi_S3_8p_gegen_Golf_VI_Gti_mit_xenon   
593    2016-04-01 21:44:15                              CLK_200__BRABUS_Optik   
660    2016-03-29 21:44:40                                       Opel_Corsa_B   
661    2016-03-18 22:49:48               Renault_Megane_Scenic_1.6_Tuev_12/17   
...                    ...                                                ...   
48882  2016-03-14 23:58:47                                         SUCHE_AUTO   
48889  2016-04-01 10:50:52                           Audi_A4_____TueV_09.2016   
48891  2016-03-20 22:50:14        MB_E_220_CDI__TAUSCHEN_nur_SUV/GELÄNDEWAGEN   
48982  2016-03-08 12:48:33                           Bmw_525D_Sport_Pro..Navi   
48985  2016-03-27 08:55:25                             Stockcar_Autos_Peugeot   
49043  2016-03-07 11:48:55                              Opel_Vectra_B_Caravan   
49082  2016-03-13 17:06:58         Passat_35i_b4_in_candy_weiss_original_lack   
49254  2016-03-21 10:49:57                           Corrado_G60_mit_2e_Motor   
49301  2016-03-20 20:52:40                           Golf_3_Variant_Kombi_1.8   
49342  2016-03-09 20:47:01  Audi_80_b4_S_Color_und_BMW_318ti_compact_Tausc...   
49350  2016-03-19 23:57:15                                     Golf_II_Diesel   
49404  2016-04-04 03:03:16                                Golf_2_BBS_LEDER_KW   
49408  2016-03-30 20:57:06                                         OPEL_ASTRA   
49479  2016-04-01 23:55:01                            Tausch_mit_frischen_TÜV   
49496  2016-03-26 13:55:28                                        Bmw_e39_520   
49504  2016-03-13 19:42:57                                 Meriva_Lpg_1_6_16v   
49507  2016-03-16 17:38:32                 BMW_528i_E39_Schaltgetriebe_Unfall   
49525  2016-03-09 17:59:02              Bmw_e

In [15]:
examhigh=autos[autos['price']>1000000]
print(examhigh.shape[0])
print(examhigh['price'].value_counts().sort_index(ascending=False))
examhigh

11
99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
Name: price, dtype: int64


date_crawled                                               name  \
2897   2016-03-12 21:50:57   Escort_MK_1_Hundeknochen_zum_umbauen_auf_RS_2000   
7814   2016-04-04 11:53:31                                        Ferrari_F40   
11137  2016-03-29 23:52:57  suche_maserati_3200_gt_Zustand_unwichtig_laufe...   
22947  2016-03-22 12:54:19                         Bmw_530d_zum_ausschlachten   
24384  2016-03-21 13:57:51                            Schlachte_Golf_3_gt_tdi   
27371  2016-03-09 15:45:47                                         Fiat_Punto   
39377  2016-03-08 23:53:51                        Tausche_volvo_v40_gegen_van   
39705  2016-03-22 14:58:27                        Tausch_gegen_gleichwertiges   
42221  2016-03-08 20:39:05                                  Leasinguebernahme   
47598  2016-03-31 18:56:54  Opel_Vectra_B_1_6i_16V_Facelift_Tuning_Showcar...   
47634  2016-04-04 21:25:21                                        Ferrari_FXX   

       seller offer_type     price  ab_test vehicle_type  registration_year  \
2897   privat    Angebot  11111111     test    limousine               1973   
7814   privat    Angebot   1300000  control        coupe               1992   
11137  privat    Angebot  10000000  control        coupe               1960   
22947  privat    Angebot   1234566  control        kombi               1999   
24384  privat    Angebot  11111111     test          NaN               1995   
27371  privat    Angebot  12345678  control          NaN               2017   
39377  privat    Angebot  12345678  control          NaN               2018   
39705  privat    Angebot  99999999  control    limousine               1999   
42221  privat    Angebot  27322222  control    limousine               2014   
47598  privat    Angebot  12345678  control    limousine               2001   
47634  privat    Angebot   3890000     test        coupe               2006   

         gearbox  powerPS     model  odometer_km  registration_month  \
2897     manuell       48    escort        50000                   3   
7814         NaN        0       NaN        50000                  12   
11137    manuell      368       NaN       100000                   1   
22947  automatik      190       NaN       150000                   2   
24384        NaN        0       NaN       150000                   0   
27371        NaN       95     punto       150000                   0   
39377    manuell       95       v40       150000                   6   
39705  automatik      224  s_klasse       150000                   9   
42221    manuell      163        c4        40000                   2   
47598    manuell      101    vectra       150000                   3   
47634        NaN      799       NaN         5000                   7   

      fuel_type           brand unrepaired_damage           ad_created  \
2897     benzin            ford              nein  2016-03-12 00:00:00   
7814        NaN  sonstige_autos              nein  2016-04-04 00:00:00   
11137    benzin  sonstige_autos              nein  2016-03-29 00:00:00   
22947    diesel             bmw               NaN  2016-03-22 00:00:00   
24384       NaN      volkswagen               NaN  2016-03-21 00:00:00   
27371       NaN            fiat               NaN  2016-03-09 00:00:00   
39377       NaN           volvo              nein  2016-03-08 00:00:00   
39705    benzin   mercedes_benz               NaN  2016-03-22 00:00:00   
42221    diesel         citroen               NaN  2016-03-08 00:00:00   
47598    benzin            opel              nein  2016-03-31 00:00:00   
47634       NaN  sonstige_autos              nein  2016-04-04 00:00:00   

       nr_pf_pictures  postal_code            last_seen  
2897                0        94469  2016-03-12 22:45:27  
7814                0        60598  2016-04-05 11:34:11  
11137               0        73033  2016-04-06 21:18:11  
22947               0        17454  2016-04-02 03:17:32  
24384               0        18519  2016-03-21 1

In [16]:
autos[autos['price'].between(1,10000000)]

date_crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
30     2016-03-14 11:47:31                         Peugeot_206_Unfallfahrzeug   
...                    ...                                                ...   
49968  2016-04-01 17:49:15                  Mercedes_Benz_190_D_2.5_Automatik   
49969  2016-03-17 18:49:02             Nissan_X_Trail_2.2_dCi_4x4_Sport_m.AHZ   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine

## Part V - Exploring the Date Columns

In [17]:
autos[['date_crawled','ad_created','last_seen']][:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In [18]:
autos['date_crawled'].str[:10].value_counts(normalize=True, dropna=False).sort_index()
#after value_counts can do .sort_index(ascending=False) , .sort_index() , .sort_values(ascending=True) and .sort_values()
#.shape[0] = 34
#.sum() = 1 to see if there it is out of 100%

2016-03-05    0.02538
2016-03-06    0.01394
2016-03-07    0.03596
2016-03-08    0.03330
2016-03-09    0.03322
2016-03-10    0.03212
2016-03-11    0.03248
2016-03-12    0.03678
2016-03-13    0.01556
2016-03-14    0.03662
2016-03-15    0.03398
2016-03-16    0.02950
2016-03-17    0.03152
2016-03-18    0.01306
2016-03-19    0.03490
2016-03-20    0.03782
2016-03-21    0.03752
2016-03-22    0.03294
2016-03-23    0.03238
2016-03-24    0.02910
2016-03-25    0.03174
2016-03-26    0.03248
2016-03-27    0.03104
2016-03-28    0.03484
2016-03-29    0.03418
2016-03-30    0.03362
2016-03-31    0.03192
2016-04-01    0.03380
2016-04-02    0.03540
2016-04-03    0.03868
2016-04-04    0.03652
2016-04-05    0.01310
2016-04-06    0.00318
2016-04-07    0.00142
Name: date_crawled, dtype: float64

- first crawled meaning: 'This is the date that the link was first located and indexed within the Fresh or Historic Indexes.'
- crawled meaning: 'A web crawler is an internet bot that systematically browses world wide web for the purpose of extracting useful information'
- All date crawled are daily from March 5th to April 7th of 2016
- No date's frequency is above 3.9% and there are 34 dates that make up the distribution
- It appears that the frequencies are around the same range as more than half of the values are 3-4%

In [19]:
autos['ad_created'].str[:10].value_counts(normalize=True, dropna=False).sort_values(ascending=False)
#after value_counts can do .sort_index(ascending=False) , .sort_index() , .sort_values(ascending=True) and .sort_values()
#.shape[0] = 34
#.sum() = 1 to see if there it is out of 100%

2016-04-03    0.03892
2016-03-20    0.03786
2016-03-21    0.03772
2016-04-04    0.03688
2016-03-12    0.03662
2016-03-14    0.03522
2016-04-02    0.03508
2016-03-28    0.03496
2016-03-07    0.03474
2016-03-29    0.03414
2016-03-19    0.03384
2016-04-01    0.03380
2016-03-15    0.03374
2016-03-30    0.03344
2016-03-08    0.03334
2016-03-09    0.03324
2016-03-22    0.03280
2016-03-11    0.03278
2016-03-26    0.03256
2016-03-23    0.03218
2016-03-31    0.03192
2016-03-25    0.03188
2016-03-10    0.03186
2016-03-17    0.03120
2016-03-27    0.03090
2016-03-16    0.03000
2016-03-24    0.02908
2016-03-05    0.02304
2016-03-13    0.01692
2016-03-06    0.01512
               ...   
2016-02-21    0.00006
2016-02-26    0.00004
2016-02-14    0.00004
2016-01-10    0.00004
2016-02-05    0.00004
2016-02-18    0.00004
2016-02-20    0.00004
2016-02-02    0.00004
2016-02-09    0.00004
2016-02-24    0.00004
2016-02-22    0.00002
2016-02-17    0.00002
2016-01-03    0.00002
2015-11-10    0.00002
2016-01-14

- The dates the ads were created range 76 dates from June 2015 to April of 2016
- The frequency of the dates are from 0.002 to 3.892% thus 3 orders of magnitude 
- Most of the ads were created in March and April of 2016

In [20]:
autos['last_seen'].str[:10].value_counts(normalize=True, dropna=False).sort_index()
#after value_counts can do .sort_index(ascending=False) , .sort_index() , .sort_values(ascending=True) and .sort_values()
#.shape[0] = 34
#.sum() = 1 to see if there it is out of 100%

2016-03-05    0.00108
2016-03-06    0.00442
2016-03-07    0.00536
2016-03-08    0.00760
2016-03-09    0.00986
2016-03-10    0.01076
2016-03-11    0.01252
2016-03-12    0.02382
2016-03-13    0.00898
2016-03-14    0.01280
2016-03-15    0.01588
2016-03-16    0.01644
2016-03-17    0.02792
2016-03-18    0.00742
2016-03-19    0.01574
2016-03-20    0.02070
2016-03-21    0.02074
2016-03-22    0.02158
2016-03-23    0.01858
2016-03-24    0.01956
2016-03-25    0.01920
2016-03-26    0.01696
2016-03-27    0.01602
2016-03-28    0.02086
2016-03-29    0.02234
2016-03-30    0.02484
2016-03-31    0.02384
2016-04-01    0.02310
2016-04-02    0.02490
2016-04-03    0.02536
2016-04-04    0.02462
2016-04-05    0.12428
2016-04-06    0.22100
2016-04-07    0.13092
Name: last_seen, dtype: float64

- All last seen are daily from March 5th to April 7th of 2016 as well
- The first three dates make up the lowest date frequencies and the last three dates make up the highest date frequencies
- The percentage range is wide: 0.11% to 22.10%

In [21]:
autos['registration_year'].describe()

count    50000.000000
mean      2005.073280
std        105.712813
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

- The mean value is 2005 which is reasonable, it makes sense that in 2016 users are selling their car on ebay that is on average ~10 years old
- The standard deviation at 105 is high, which means that the registration year highly deviates from the mean 2005.
- The minimum amd maximum are unreasonable as in 1000 cars did not exist and 9999 is in the future for this dataset

## Part VI - Dealing with Incorrect Registration Year Data

- Realistically, we would check the year the ad was created and make it so that the registration year is not higher than that, however in this case, because all the ads were created in 2015-2016, 2016 will be the upper limit of the registration years included
- After some research, it appears that the earliest registration date was 1903 so that will be ther lower limit of the registration years included

In [22]:
# Calculate distribution of values
autos['registration_year'].value_counts(normalize=True)

2000    0.06708
2005    0.06030
1999    0.06000
2004    0.05474
2003    0.05454
2006    0.05416
2001    0.05406
2002    0.05066
1998    0.04906
2007    0.04608
2008    0.04462
2009    0.04196
1997    0.04056
2011    0.03268
2010    0.03194
2017    0.02906
1996    0.02888
2012    0.02646
2016    0.02632
1995    0.02626
2013    0.01612
2014    0.01332
1994    0.01320
2018    0.00984
1993    0.00890
2015    0.00798
1990    0.00790
1992    0.00782
1991    0.00712
1989    0.00362
         ...   
1950    0.00006
1955    0.00004
9000    0.00004
1954    0.00004
1800    0.00004
1957    0.00004
1941    0.00004
1951    0.00004
1934    0.00004
4100    0.00002
4800    0.00002
1953    0.00002
1111    0.00002
1927    0.00002
6200    0.00002
4500    0.00002
1943    0.00002
5911    0.00002
1939    0.00002
1938    0.00002
2800    0.00002
8888    0.00002
1000    0.00002
1500    0.00002
1948    0.00002
1931    0.00002
1929    0.00002
1001    0.00002
9996    0.00002
1952    0.00002
Name: registration_year,

In [23]:
# Percentage of car listings that fall outside the 1900 to 2016
# ~ = NOT
(~autos['registration_year'].between(1903,2016)).sum()/autos.shape[0]*100

3.9440000000000004

In [24]:
# delete values that are not between 1903 and 2016
autos=autos[autos['registration_year'].between(1903,2016)]
autos.shape

(48028, 20)

In [25]:
# Recalculate distribution of remaining values
autos['registration_year'].value_counts(normalize=True)

2000    0.069834
2005    0.062776
1999    0.062464
2004    0.056988
2003    0.056779
2006    0.056384
2001    0.056280
2002    0.052740
1998    0.051074
2007    0.047972
2008    0.046452
2009    0.043683
1997    0.042225
2011    0.034022
2010    0.033251
1996    0.030066
2012    0.027546
2016    0.027401
1995    0.027338
2013    0.016782
2014    0.013867
1994    0.013742
1993    0.009265
2015    0.008308
1990    0.008224
1992    0.008141
1991    0.007412
1989    0.003769
1988    0.002957
1985    0.002186
          ...   
1966    0.000458
1977    0.000458
1975    0.000396
1969    0.000396
1965    0.000354
1964    0.000250
1910    0.000187
1963    0.000187
1959    0.000146
1961    0.000125
1956    0.000104
1958    0.000083
1937    0.000083
1962    0.000083
1950    0.000062
1954    0.000042
1941    0.000042
1934    0.000042
1957    0.000042
1951    0.000042
1955    0.000042
1931    0.000021
1953    0.000021
1943    0.000021
1938    0.000021
1939    0.000021
1927    0.000021
1929    0.0000

- frequencies increased for all remaining values because the there are less values involved
- most values are 1985 and above

## Part VII - Exploring Price by Brand

In [26]:
all_brands=autos['brand'].value_counts(normalize=True)[:20]
print(all_brands)
print(all_brands.shape[0])
print(all_brands.sum())
print(all_brands.index)

volkswagen        0.212126
bmw               0.110019
opel              0.108166
mercedes_benz     0.095361
audi              0.086387
ford              0.069793
renault           0.047347
peugeot           0.029524
fiat              0.025860
seat              0.018177
skoda             0.016032
mazda             0.015137
nissan            0.015095
citroen           0.013929
smart             0.013909
toyota            0.012472
sonstige_autos    0.010952
hyundai           0.009848
volvo             0.009245
mini              0.008641
Name: brand, dtype: float64
20
0.9280211543266427
Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat', 'skoda', 'mazda', 'nissan', 'citroen',
       'smart', 'toyota', 'sonstige_autos', 'hyundai', 'volvo', 'mini'],
      dtype='object')


In [27]:
above_this=autos['brand'].value_counts(normalize=True)
above_this=above_this[above_this>0.05]
print(above_this)
print(above_this.shape[0])
print(above_this.sum())
print(above_this.index)

volkswagen       0.212126
bmw              0.110019
opel             0.108166
mercedes_benz    0.095361
audi             0.086387
ford             0.069793
Name: brand, dtype: float64
6
0.6818522528525027
Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')


- The brands that make up the top 20 brands come from a varieties of countries and classes
- The top 20 brands make up almost 93% of the brands while the top 5 make up about 68% of the brands. Because the top 20 covers a significantly higher percentage of the listings

In [28]:
aggdat={}
for brand in all_brands.index:
    bmean=autos['price'][autos['brand'] == brand].mean()
    aggdat[brand]=bmean
print(aggdat)
#sort dictionary by values https://stackoverflow.com/questions/613183/how-do-i-sort-a-dictionary-by-value

{'mini': 10460.012048192772, 'nissan': 4664.891034482758, 'seat': 4296.492554410081, 'mazda': 4010.7716643741405, 'mercedes_benz': 30317.447816593885, 'ford': 7263.015811455847, 'fiat': 2711.8011272141707, 'citroen': 44534.79671150971, 'hyundai': 5308.53911205074, 'renault': 2395.4164467897976, 'smart': 3542.706586826347, 'peugeot': 3039.4682651622, 'audi': 9093.65003615329, 'bmw': 8334.645155185466, 'skoda': 6334.91948051948, 'volkswagen': 6516.457597173145, 'toyota': 5115.33388981636, 'volvo': 4757.108108108108, 'sonstige_autos': 39621.77946768061, 'opel': 5252.61655437921}


- Most brand are within a range of \$2,000 to \$10,000 while others are more expensive and around \$30,000 - \$40,000

## Part VIII - Storing Aggregate Data in a DataFrame

In [29]:
miledat={}
for brand in all_brands.index:
    mmean=autos['odometer_km'][autos['brand'] == brand].mean()
    miledat[brand]=mmean
print(miledat)

{'mini': 88602.40963855422, 'nissan': 118572.41379310345, 'seat': 121563.57388316152, 'mazda': 124745.5295735901, 'mercedes_benz': 130860.26200873363, 'ford': 124046.83770883054, 'fiat': 116553.94524959743, 'citroen': 119461.88340807175, 'hyundai': 106511.62790697675, 'renault': 128183.81706244503, 'smart': 99595.80838323354, 'peugeot': 127136.81241184767, 'audi': 129287.78018799711, 'bmw': 132434.70855412565, 'skoda': 110954.54545454546, 'volkswagen': 128730.36906164115, 'toyota': 115709.51585976628, 'volvo': 138355.85585585586, 'sonstige_autos': 87262.35741444866, 'opel': 129227.14148219442}


In [36]:
pd_aggdat=pd.Series(aggdat)
pd_miledat=pd.Series(miledat)
df_dat=pd.DataFrame(pd_aggdat, columns=['mean_price'])
df_dat['mean_milage']=pd_miledat
df_dat

mean_price    mean_milage
audi             9093.650036  129287.780188
bmw              8334.645155  132434.708554
citroen         44534.796712  119461.883408
fiat             2711.801127  116553.945250
ford             7263.015811  124046.837709
hyundai          5308.539112  106511.627907
mazda            4010.771664  124745.529574
mercedes_benz   30317.447817  130860.262009
mini            10460.012048   88602.409639
nissan           4664.891034  118572.413793
opel             5252.616554  129227.141482
peugeot          3039.468265  127136.812412
renault          2395.416447  128183.817062
seat             4296.492554  121563.573883
skoda            6334.919481  110954.545455
smart            3542.706587   99595.808383
sonstige_autos  39621.779468   87262.357414
toyota           5115.333890  115709.515860
volkswagen       6516.457597  128730.369062
volvo            4757.108108  138355.855856

In [37]:
df_dat.describe()

mean_price    mean_milage
count     20.000000      20.000000
mean   10378.593473  118889.859745
std    12377.297192   14071.228575
min     2395.416447   87262.357414
25%     4225.062332  114520.773258
50%     5280.577833  122805.205796
75%     8524.396375  128854.562167
max    44534.796712  138355.855856

In [45]:
df_dat.sort_values(by='mean_price')

mean_price    mean_milage
renault          2395.416447  128183.817062
fiat             2711.801127  116553.945250
peugeot          3039.468265  127136.812412
smart            3542.706587   99595.808383
mazda            4010.771664  124745.529574
seat             4296.492554  121563.573883
nissan           4664.891034  118572.413793
volvo            4757.108108  138355.855856
toyota           5115.333890  115709.515860
opel             5252.616554  129227.141482
hyundai          5308.539112  106511.627907
skoda            6334.919481  110954.545455
volkswagen       6516.457597  128730.369062
ford             7263.015811  124046.837709
bmw              8334.645155  132434.708554
audi             9093.650036  129287.780188
mini            10460.012048   88602.409639
mercedes_benz   30317.447817  130860.262009
sonstige_autos  39621.779468   87262.357414
citroen         44534.796712  119461.883408

In [46]:
df_dat.sort_values(by='mean_milage')

mean_price    mean_milage
sonstige_autos  39621.779468   87262.357414
mini            10460.012048   88602.409639
smart            3542.706587   99595.808383
hyundai          5308.539112  106511.627907
skoda            6334.919481  110954.545455
toyota           5115.333890  115709.515860
fiat             2711.801127  116553.945250
nissan           4664.891034  118572.413793
citroen         44534.796712  119461.883408
seat             4296.492554  121563.573883
ford             7263.015811  124046.837709
mazda            4010.771664  124745.529574
peugeot          3039.468265  127136.812412
renault          2395.416447  128183.817062
volkswagen       6516.457597  128730.369062
opel             5252.616554  129227.141482
audi             9093.650036  129287.780188
mercedes_benz   30317.447817  130860.262009
bmw              8334.645155  132434.708554
volvo            4757.108108  138355.855856

- After sorting values by both mean price and mean milage it is safe to say that amongst the top 20 most popular brands there is weak correlation between the mean price and mean milage